client/lib/iothub/iothubClient.ts

```diff
@@ -0,0 +1,88 @@
+import { IotHubConnectionConfig } from "./iotHubConnectionConfig";
+import { ConnectionContext, ConnectionContextOptions } from "../connectionContext";
+import { IotSasTokenProvider } from "./iotSas";
+import * as debugModule from "debug";
+import { translate } from "../errors";
+import { closeConnection } from "../rhea-promise";
+const debug = debugModule("azure:event-hubs:iothubClient");
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+export class IotHubClient {
+  /**
+   * @property {string} connectionString the IotHub connection string.
+   */
+  connectionString: string;
+
+  constructor(connectionString: string) {
+    this.connectionString = connectionString;
+  }
+  /**
+   * Constructs the EventHub connection string by catching the redirect error and parsing the error
+   * information.
+   * @param {ConnectionContextOptions} [options] optional parameters to be provided while creating
+   * the connection context.
+   * @return {Promise<string>} Promise<string>
+   */
+  async getEventHubConnectionString(options?: ConnectionContextOptions): Promise<string> {
+    const iothubconfig = IotHubConnectionConfig.create(this.connectionString);
+    const config = IotHubConnectionConfig.convertToConnectionConfig(iothubconfig);
+    let result = `SharedAccessKeyName=${config.sharedAccessKeyName};SharedAccessKey=${config.sharedAccessKey}`;
+    if (!options) options = {};
+    options.tokenProvider = new IotSasTokenProvider(config.endpoint, config.sharedAccessKeyName, config.sharedAccessKey);
+    options.managementSessionAddress = `/messages/events/$management`;
+    const context = ConnectionContext.create(config, options);
+    try {
+      debug("Getting the hub runtime info from the iothub connection string to get the redirect error.");
+      await context.managementSession!.getHubRuntimeInformation();
+    } catch (err) {
+      const error = translate(err);
+      debug("IotHubClient received the error: %O", error);
+      if (error.name !== "LinkRedirectError" || !error.info) {
+        throw error;
+      }
+      if (!error.info.hostname || !error.info.address) {
+        const msg = `The received redirect error from IotHub is malformed. ${error.stack}\n${error.info}`;
+        throw new Error(msg);
+      }
+      result = `Endpoint=sb://${error.info.hostname}/;` + result;
```

Comment:
these string joins feel brittle. it might be more robust to have something like:

```ts
const parts = new Map();
parts.set('Endpoint', `sb://${error.info.hostname}/`);
// ... and so on
const result = Array.from(parts).map(part => `${part[0]}=${part[1]}`).join(';')
```